HUANG Flora

#  Tableau 2 : Résultats d'efficacité et de sécurité

Consigne : Élaborer un tableau de résultats présentant :
- Le critère de jugement principal avec analyse statistique complète
- Estimations d'effet (différence de proportions) avec intervalles de confiance à 95%
- Valeurs de p et significativité statistique


# Proposition de plan : 

Afin d'élaborer notre tableau 2 et démontrer si les résultats sont efficaces et sécurisable, il est nécessaire qu'un patient satisfasse tous les critères de jugement principal. 

Pour cela, un patient donné doit : 
- toujours être présent au 13ème jour de l'étude. (**Rétention**)
- son test urinaire au 13ème jour est négatif aux opioïdes. (**Abstinence**)
  
Ensuite, une analyse statistique est effectuée pour comparer l'efficacité des 2 traitement mais aussi pour compléter le tableau.
> Le test de Khi-2 est appliqué: permet de faire une compaison de nos 2 traitements et observer leurs efficaciés (BUP/NAL et CLO) 


Nous commençons par effectuer une recherche (Ctrl F) du mot "urine" dans le fichier "CTN001 ou CNT002.<br>

Cela nous conduit aux pages annotées par "DOMAIN: L qui nous renvoi e, un fichier nommé "lb.csv".

# Analyse de la rétention et de l'abstinence

In [1]:
#Chargement des données et bibliothèques
import numpy as np # manipuler nos données numérique 
import polars as pl # manipuler nos données tabulaire 
import scipy as sp #faire une analyse scientifique et statistique 

In [2]:
lb1 = pl.read_csv('lb1.csv')
lb2 = pl.read_csv('lb2.csv')

In [3]:
print(lb1.shape)
print(lb2.shape)

(6332, 27)
(11645, 27)


In [4]:
lb = pl.concat([lb1, lb2]) 
lb.sample(3)

STUDYID,DOMAIN,USUBJID,EPOCH,LBSEQ,LBTESTCD,LBTEST,LBCAT,LBORRES,LBORRESU,LBORNRLO,LBORNRHI,LBSTRESC,LBSTRESN,LBSTRESU,LBSTNRLO,LBSTNRHI,LBNRIND,LBSTAT,LBREASND,LBSPEC,LBMETHOD,LBBLFL,VISIT,VISITNUM,LBDTC,LBDY
str,str,str,str,i64,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,i64,i64,str
"""NIDA-CTN-0002""","""LB""","""02_048763""","""ACTIVE""",34,"""NONE""","""NONE""","""DRUG SCREEN""",null,null,null,null,null,""" """,null,""" """,""" """,null,"""NOT DONE""","""URINE NOT COLLECTED (SPECIFY I…","""URINE""","""CENTRAL LAB""",null,"""STUDY DAY 14""",14,2001,"""14"""
"""NIDA-CTN-0001""","""LB""","""01_038661""","""ACTIVE""",20,"""COCAINE""","""COCAINE""","""DRUG SCREEN""","""NEGATIVE""",null,null,null,"""NEGATIVE""",""" """,null,""" """,""" """,null,null,null,"""URINE""","""CENTRAL LAB""",null,"""STUDY DAY 3""",3,2002,"""3"""
"""NIDA-CTN-0002""","""LB""","""02_078138""","""ACTIVE""",35,"""CREAT""","""CREATININE""","""CHEMISTRY""","""45.7""","""mg/dL""",null,null,"""45.7""","""45.7""","""mg/dL""",""" """,""" """,null,null,null,"""URINE""","""CENTRAL LAB""",null,"""STUDY DAY 10""",10,2001,"""10"""


In [5]:
lb.shape

(17977, 27)

In [6]:
# Sélectionner les colonnes qui nous intéresses
lb = lb.select(
    pl.col('USUBJID'), #identifiant
    pl.col('LBTESTCD'), # type de médicamment
    pl.col('LBORRES'), # test urinaire de la positivité
    pl.col('LBSTRESC'), # test urinaire de la positivité 
    pl.col('LBSPEC'), # type de test
    pl.col('VISIT'), # jour de visite
    pl.col('VISITNUM'), # jour de visite
)
lb

USUBJID,LBTESTCD,LBORRES,LBSTRESC,LBSPEC,VISIT,VISITNUM
str,str,str,str,str,str,i64
"""01_000579""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_000579""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_000579""","""BARBIT""","""POSITIVE""","""POSITIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_000579""","""BENZOD""","""POSITIVE""","""POSITIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_000579""","""BENZOD""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
…,…,…,…,…,…,…
"""02_099926""","""THC""","""NEGATIVE""","""NEGATIVE""","""URINE""","""STUDY DAY 3""",3
"""02_099926""","""THC""","""NEGATIVE""","""NEGATIVE""","""URINE""","""STUDY DAY 8""",8
"""02_099926""","""THC""","""NEGATIVE""","""NEGATIVE""","""URINE""","""STUDY DAY 11""",11


Exercie proposé : Vérifier que le numéro de visite i (colonne "VISITNUM") correspond bien au i-ème jour de l'étude/traitement (colonne "VISIT" -- quand c'est renseigné).

In [7]:
lb_exo = lb.select(
    pl.col('VISIT'),
    pl.col('VISITNUM'),
)
lb_exo

VISIT,VISITNUM
str,i64
"""BASELINE, ANY DAY PRIOR TO FIR…",0
"""BASELINE, ANY DAY PRIOR TO FIR…",0
"""BASELINE, ANY DAY PRIOR TO FIR…",0
"""BASELINE, ANY DAY PRIOR TO FIR…",0
"""BASELINE, ANY DAY PRIOR TO FIR…",0
…,…
"""STUDY DAY 3""",3
"""STUDY DAY 8""",8
"""STUDY DAY 11""",11


In [8]:
lb.select(pl.col('VISIT').unique())

VISIT
str
"""STUDY DAY 8"""
"""FOLLOW-UP 3"""
"""STUDY DAY 14"""
"""UNKNOWN"""
"""STUDY DAY 10"""
…
"""STUDY DAY 2"""
"""BASELINE, ANY DAY PRIOR TO FIR…"
"""STUDY DAY 11"""


In [9]:
lb.select(pl.col('VISITNUM').unique())

VISITNUM
i64
-99
0
2
3
4
…
13
14
15


Nous prenons en compte les follow-up, qui correspondent à la continuité des visites après la phase de traitement.
Nous les transformons donc en visites correspondantes selon la logique suivante :
- FOLLOW-UP 1 = 15 (pour un suivie de 1 mois)
- FOLLOW-UP 2 = 16 (pour un suivi de 3 mois)
- FOLLOW-UP 3 = 17 (pour un suivi de 9 mois)

In [10]:
lb_exo = lb_exo.with_columns([
    pl.when(pl.col("VISIT").str.contains("FOLLOW-UP 1"))
        .then(15)
    .when(pl.col("VISIT").str.contains("FOLLOW-UP 2"))
        .then(16)
    .when(pl.col("VISIT").str.contains("FOLLOW-UP 3"))
        .then(17)
    .otherwise(
        pl.col("VISIT").str.extract(r"(\d+)").cast(pl.Int64)
    )
    .alias("VISIT_EXTRACT")
])

lb_exo = lb_exo.with_columns(
    (pl.col("VISIT_EXTRACT") == pl.col("VISITNUM")).alias("equal")
)

lb_exo

VISIT,VISITNUM,VISIT_EXTRACT,equal
str,i64,i64,bool
"""BASELINE, ANY DAY PRIOR TO FIR…",0,null,null
"""BASELINE, ANY DAY PRIOR TO FIR…",0,null,null
"""BASELINE, ANY DAY PRIOR TO FIR…",0,null,null
"""BASELINE, ANY DAY PRIOR TO FIR…",0,null,null
"""BASELINE, ANY DAY PRIOR TO FIR…",0,null,null
…,…,…,…
"""STUDY DAY 3""",3,3,true
"""STUDY DAY 8""",8,8,true
"""STUDY DAY 11""",11,11,true


In [11]:
lb_exo = lb_exo.select(
    pl.col("equal").value_counts()
)
lb_exo

equal
struct[2]
"{null,6552}"
"{true,11425}"


total equal : 11425 + 6552 = 17977 réponses

Ainsi, nous observons que le numéro de visite i (colonne VISITNUM) correspond bien au i-ème jour de l’étude ou du traitement (colonne VISIT).
Nous remarquons également la valeur -99 dans la colonne VISITNUM, qui correspond à "UNKNOWN" dans la colonne VISIT. Nous pouvons supposer qu’il s’agit d’une valeur erronée.

Par la suite pour déterminer la test urinaire des patients, il suffit de fitrer les colonnes 'LBORRES' et/ou 'LBSTRESC' 

Vérifions d'abord les valeurs prises par 'LBSPEC' pour les différents patients.

In [12]:
lb.select(
    pl.col('LBSPEC'),
).unique()

LBSPEC
str
"""URINE"""


In [13]:
lb.select(
    pl.col('LBSPEC').unique_counts()
)

LBSPEC
u32
17977


Nous obtenons 17977 fois la valeur 'URINE' qui correspond au nombre de départ (fusion des 2 fichiers lb1 et lb2)

In [14]:
lb.select(
    pl.col('USUBJID'),
).unique()

USUBJID
str
"""02_048837"""
"""02_085773"""
"""02_044884"""
"""02_090309"""
"""02_090484"""
…
"""02_053581"""
"""02_097455"""
"""02_086880"""


In [15]:
lb.select(
    pl.col('USUBJID').unique_counts()
)

USUBJID
u32
21
43
43
56
58
…
29
49
14


A l'ocurrance, nous avons bien 405 tests urianires pour 405 patients.

# Rétention 
Nous conservons uniquement les participants présents jusqu’à la fin de l'étude : permet d'évaluer l'acceptabilité du traitement, sa fiabilité et son efficacité.

In [16]:
lb_ret = lb.filter(
    pl.col('VISITNUM') > 12,
)
lb_ret

USUBJID,LBTESTCD,LBORRES,LBSTRESC,LBSPEC,VISIT,VISITNUM
str,str,str,str,str,str,i64
"""01_000579""","""NONE""",null,null,"""URINE""","""FOLLOW-UP 1""",15
"""01_000579""","""NONE""",null,null,"""URINE""","""FOLLOW-UP 2""",16
"""01_000579""","""NONE""",null,null,"""URINE""","""FOLLOW-UP 3""",17
"""01_001362""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""FOLLOW-UP 1""",15
"""01_001362""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""FOLLOW-UP 2""",16
…,…,…,…,…,…,…
"""02_099926""","""NONE""",null,null,"""URINE""","""FOLLOW-UP 1""",15
"""02_099926""","""OPIATE""","""POSITIVE""","""POSITIVE""","""URINE""","""FOLLOW-UP 2""",16
"""02_099926""","""OPIATE""","""NEGATIVE""","""NEGATIVE""","""URINE""","""FOLLOW-UP 3""",17


In [17]:
lb_ret.select(
    pl.col('VISITNUM'),
).unique()

VISITNUM
i64
15
13
16
14
17


Si un patient apparaît pour un numéro de visite ("VISITNUM") égal à 13 ou plus, nous considérons qu'il satisfait au "sous-critère" de rétention. 

In [18]:
lb_ret.select(
    pl.col('USUBJID').unique_counts()
)

USUBJID
u32
3
15
4
20
20
…
3
17
3


En filtrant les patients, il ne reste plus que 324 participants sur 405 (valeur déterminée précedamment) qui valide les critères de jugement principals.
Nous pouvons supposer qu'entre temps des patients ont arrêté l'expérience...

Il est nécessaire par la suite de les trier selon le bras de traitement afin de connaître leurs efficacités et faire une comparaison. Pour cela, nous effectuons une jointure avec la table contenant les données démographiques, soit le fichier "dm.csv".

In [19]:
dm1 = pl.read_csv('dm1.csv')
dm2 = pl.read_csv('dm2.csv')
dm = pl.concat([dm1, dm2])

dm.sample(4)

STUDYID,DOMAIN,USUBJID,EPOCH,VISIT,VISITNUM,RFSTDTC,RFENDTC,SITEID,BRTHDTC,AGE,AGEU,SEX,RACE,ETHNIC,ARMCD,ARM,COUNTRY,DMDTC,DMDY
str,str,str,str,str,i64,i64,i64,str,i64,str,str,str,str,str,str,str,str,i64,str
"""NIDA-CTN-0001""","""DM""","""01_003653""","""SCREENING""","""BASELINE, ANY DAY PRIOR TO FIR…",0,2002,2002,null,1975,"""26.650239562""","""YEARS""","""M""","""WHITE""",null,"""BUPNAL""","""BUPRENORPHINE/NALOXONE""","""USA""",2002,"""-5"""
"""NIDA-CTN-0002""","""DM""","""02_099926""","""SCREENING""","""BASELINE, ANY DAY PRIOR TO FIR…",0,2001,2001,null,1970,"""30.726899384""","""YEARS""","""M""","""WHITE""",null,"""BUPNAL""","""BUPRENORPHINE/NALOXONE""","""USA""",2001,"""-1"""
"""NIDA-CTN-0001""","""DM""","""01_076020""","""SCREENING""","""BASELINE, ANY DAY PRIOR TO FIR…",0,2001,2001,null,1958,"""42.929500342""","""YEARS""","""M""","""WHITE""",null,"""BUPNAL""","""BUPRENORPHINE/NALOXONE""","""USA""",2001,"""-5"""
"""NIDA-CTN-0001""","""DM""","""01_085459""","""SCREENING""","""BASELINE, ANY DAY PRIOR TO FIR…",0,2001,2002,null,1965,"""35.967145791""","""YEARS""","""F""","""WHITE""",null,"""BUPNAL""","""BUPRENORPHINE/NALOXONE""","""USA""",2001,"""-6"""


In [20]:
dm.select(
    pl.col('USUBJID'),
    pl.col('ARMCD'),
)

USUBJID,ARMCD
str,str
"""01_000579""","""CLON"""
"""01_001362""","""CLON"""
"""01_001490""","""BUPNAL"""
"""01_002199""","""BUPNAL"""
"""01_002844""","""BUPNAL"""
…,…
"""02_098425""","""CLON"""
"""02_098689""","""SCRFAIL"""
"""02_099053""","""BUPNAL"""


In [21]:
dm.shape

(411, 20)

In [22]:
dm.select(
    pl.col('USUBJID'),
).unique()

USUBJID
str
"""01_073890"""
"""02_086880"""
"""02_080480"""
"""02_080936"""
"""02_024208"""
…
"""02_048803"""
"""02_009190"""
"""02_084318"""


In [23]:
#Calcule des effectifs de départ :
n0 = dm.group_by(
    pl.col('ARMCD'),
    maintain_order=True,
).agg(
    pl.col('USUBJID').unique_counts().len()
)

n0

ARMCD,USUBJID
str,u32
"""CLON""",110
"""BUPNAL""",233
"""SCRFAIL""",68


Nous retrouvons bien 411 identifiants uniques.

In [24]:
#Jointure 
lb_ret = lb_ret.join(
    dm,
    on='USUBJID',
)
lb_ret

USUBJID,LBTESTCD,LBORRES,LBSTRESC,LBSPEC,VISIT,VISITNUM,STUDYID,DOMAIN,EPOCH,VISIT_right,VISITNUM_right,RFSTDTC,RFENDTC,SITEID,BRTHDTC,AGE,AGEU,SEX,RACE,ETHNIC,ARMCD,ARM,COUNTRY,DMDTC,DMDY
str,str,str,str,str,str,i64,str,str,str,str,i64,i64,i64,str,i64,str,str,str,str,str,str,str,str,i64,str
"""01_000579""","""NONE""",null,null,"""URINE""","""FOLLOW-UP 1""",15,"""NIDA-CTN-0001""","""DM""","""SCREENING""","""BASELINE, ANY DAY PRIOR TO FIR…",0,2001,2001,null,1974,"""27.282683094""","""YEARS""","""F""","""WHITE""",null,"""CLON""","""CLONIDINE""","""USA""",2001,"""-1"""
"""01_000579""","""NONE""",null,null,"""URINE""","""FOLLOW-UP 2""",16,"""NIDA-CTN-0001""","""DM""","""SCREENING""","""BASELINE, ANY DAY PRIOR TO FIR…",0,2001,2001,null,1974,"""27.282683094""","""YEARS""","""F""","""WHITE""",null,"""CLON""","""CLONIDINE""","""USA""",2001,"""-1"""
"""01_000579""","""NONE""",null,null,"""URINE""","""FOLLOW-UP 3""",17,"""NIDA-CTN-0001""","""DM""","""SCREENING""","""BASELINE, ANY DAY PRIOR TO FIR…",0,2001,2001,null,1974,"""27.282683094""","""YEARS""","""F""","""WHITE""",null,"""CLON""","""CLONIDINE""","""USA""",2001,"""-1"""
"""01_001362""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""FOLLOW-UP 1""",15,"""NIDA-CTN-0001""","""DM""","""SCREENING""","""BASELINE, ANY DAY PRIOR TO FIR…",0,2001,2001,null,1960,"""41.429158111""","""YEARS""","""F""","""BLACK, AFRICAN AMERICAN, OR NE…",null,"""CLON""","""CLONIDINE""","""USA""",2001,"""1"""
"""01_001362""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""FOLLOW-UP 2""",16,"""NIDA-CTN-0001""","""DM""","""SCREENING""","""BASELINE, ANY DAY PRIOR TO FIR…",0,2001,2001,null,1960,"""41.429158111""","""YEARS""","""F""","""BLACK, AFRICAN AMERICAN, OR NE…",null,"""CLON""","""CLONIDINE""","""USA""",2001,"""1"""
…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…
"""02_099926""","""NONE""",null,null,"""URINE""","""FOLLOW-UP 1""",15,"""NIDA-CTN-0002""","""DM""","""SCREENING""","""BASELINE, ANY DAY PRIOR TO FIR…",0,2001,2001,null,1970,"""30.726899384""","""YEARS""","""M""","""WHITE""",null,"""BUPNAL""","""BUPRENORPHINE/NALOXONE""","""USA""",2001,"""-1"""
"""02_099926""","""OPIATE""","""POSITIVE""","""POSITIVE""","""URINE""","""FOLLOW-UP 2""",16,"""NIDA-CTN-0002""","""DM""","""SCREENING""","""BASELINE, ANY DAY PRIOR TO FIR…",0,2001,2001,null,1970,"""30.726899384""","""YEARS""","""M""","""WHITE""",null,"""BUPNAL""","""BUPRENORPHINE/NALOXONE""","""USA""",2001,"""-1"""
"""02_099926""","""OPIATE""","""NEGATIVE""","""NEGATIVE""","""URINE""","""FOLLOW-UP 3""",17,"""NIDA-CTN-0002""","""DM""","""SCREENING""","""BASELINE, ANY DAY PRIOR TO FIR…",0,2001,2001,null,1970,"""30.726899384""","""YEARS""","""M""","""WHITE""",null,"""BUPNAL""","""BUPRENORPHINE/NALOXONE""","""USA""",2001,"""-1"""


In [25]:
# Colonnes qui nous intéresse 
lb_ret = lb_ret.select(
    pl.col("USUBJID"),
    pl.col("VISIT"),
    pl.col("VISITNUM"),
    pl.col("EPOCH"),
    pl.col("ARMCD"),
    pl.col("LBORRES"),
    pl.col("LBSTRESC"),
)
lb_ret

USUBJID,VISIT,VISITNUM,EPOCH,ARMCD,LBORRES,LBSTRESC
str,str,i64,str,str,str,str
"""01_000579""","""FOLLOW-UP 1""",15,"""SCREENING""","""CLON""",null,null
"""01_000579""","""FOLLOW-UP 2""",16,"""SCREENING""","""CLON""",null,null
"""01_000579""","""FOLLOW-UP 3""",17,"""SCREENING""","""CLON""",null,null
"""01_001362""","""FOLLOW-UP 1""",15,"""SCREENING""","""CLON""","""NEGATIVE""","""NEGATIVE"""
"""01_001362""","""FOLLOW-UP 2""",16,"""SCREENING""","""CLON""","""NEGATIVE""","""NEGATIVE"""
…,…,…,…,…,…,…
"""02_099926""","""FOLLOW-UP 1""",15,"""SCREENING""","""BUPNAL""",null,null
"""02_099926""","""FOLLOW-UP 2""",16,"""SCREENING""","""BUPNAL""","""POSITIVE""","""POSITIVE"""
"""02_099926""","""FOLLOW-UP 3""",17,"""SCREENING""","""BUPNAL""","""NEGATIVE""","""NEGATIVE"""


In [26]:
# Vérification des identifiants uniques > 12 
lb_ret.select(
    pl.col('USUBJID'),
).unique()

USUBJID
str
"""02_080478"""
"""01_050792"""
"""02_062244"""
"""02_043183"""
"""02_002504"""
…
"""01_055880"""
"""02_021463"""
"""02_087009"""


Exercice proposé : Vérifier que la colonne "LBSTRESC" est redondante par rapport à la colonne "LBORRES".

In [30]:
n_redon = lb_ret.select(
    pl.col('LBORRES'), 
    pl.col('LBSTRESC'),
)

# Ajoute de la colonne 'equal'
n_redon = n_redon.with_columns(
    (pl.col("LBORRES") == pl.col("LBSTRESC")).alias("equal")
)
n_redon

LBORRES,LBSTRESC,equal
str,str,bool
null,null,null
null,null,null
null,null,null
"""NEGATIVE""","""NEGATIVE""",true
"""NEGATIVE""","""NEGATIVE""",true
…,…,…
null,null,null
"""POSITIVE""","""POSITIVE""",true
"""NEGATIVE""","""NEGATIVE""",true


In [31]:
n_redon_toto = n_redon.select(
    pl.col("equal").value_counts()
)
n_redon_toto

equal
struct[2]
"{true,5777}"
"{null,525}"


5777 + 525 = 6302

La colonne "LBSTRESC" est bien redondante par rapport à la colonne "LBORRES".

In [32]:
# Nombre de participants selon le bras de traitement et après les traitements 
n = lb_ret.group_by(
    pl.col('ARMCD'),
    maintain_order=True,
).agg(
    pl.col('USUBJID').unique_counts().len()
)

n

ARMCD,USUBJID
str,u32
"""CLON""",96
"""BUPNAL""",227
"""SCRFAIL""",1


Pour le calcul de l’abstinence, nous partons du principe qu’un patient respectant l’ensemble des critères de jugement est un patient présent tout au long de l’étude, c’est-à-dire jusqu’au 13ᵉ jour, et dont les tests urinaires sont négatifs pendant toute cette période.

Par exemple, si un patient présente un test positif au 9ᵉ jour puis un test négatif au 13ᵉ jour, il n’est pas pris en compte comme abstinent.
De plus, les patients ayant au moins une valeur manquante susceptible de masquer un résultat positif ne sont pas inclus dans l’analyse.

Dans cette analyse, l’abstinence est évaluée à un moment précis : après 12 jours de traitement, sur la base du test urinaire réalisé au 13ᵉ jour. Les données de suivi ultérieur (follow-up), correspondant aux visites postérieures au 13ᵉ jour, ne sont donc pas prises en compte.

In [33]:
# Filtrer les patients présents et effectuant leurs visites les 13 jours
lb_13 = lb.filter(
    pl.col("VISITNUM") <= 13
)
lb_13.shape

(12778, 7)

In [45]:
# Supprimer les patients ayant au moins une valeur manquante durant la phase d'expérimentation 
lb_13 = lb_13.filter(
    pl.col("LBORRES").is_not_null()
)
lb_13

USUBJID,LBTESTCD,LBORRES,LBSTRESC,LBSPEC,VISIT,VISITNUM
str,str,str,str,str,str,i64
"""01_000579""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_000579""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_000579""","""BARBIT""","""POSITIVE""","""POSITIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_000579""","""BENZOD""","""POSITIVE""","""POSITIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_000579""","""BENZOD""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
…,…,…,…,…,…,…
"""02_099926""","""TCA""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""02_099926""","""THC""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""02_099926""","""THC""","""NEGATIVE""","""NEGATIVE""","""URINE""","""STUDY DAY 3""",3


In [49]:
# Identifier les patients ayant au moins un test POSITIVE
patients_pos = (
    lb_13.filter(pl.col("LBORRES") == "POSITIVE")
      .select("USUBJID")
      .unique()
)

In [50]:
# Supprimer tous les enregistrements correspondant à ces patients.
lb_sans_positive = lb_13.join(
    patients_pos,
    on="USUBJID",
    how="anti"   # anti-join = exclut ces patients
)

lb_sans_positive

USUBJID,LBTESTCD,LBORRES,LBSTRESC,LBSPEC,VISIT,VISITNUM
str,str,str,str,str,str,i64
"""01_009835""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_009835""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_009835""","""BARBIT""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_009835""","""BENZOD""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_009835""","""BENZOD""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
…,…,…,…,…,…,…
"""02_074821""","""METHADON""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""02_074821""","""MORPHINE""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""02_074821""","""PCP""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0


In [47]:
# Filtrer les patients ayant des tests négatifs 
abs_neg = lb_sans_positive.filter(pl.col("LBORRES") == "NEGATIVE")
abs_neg

USUBJID,LBTESTCD,LBORRES,LBSTRESC,LBSPEC,VISIT,VISITNUM
str,str,str,str,str,str,i64
"""01_009835""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_009835""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_009835""","""BARBIT""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_009835""","""BENZOD""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_009835""","""BENZOD""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
…,…,…,…,…,…,…
"""02_074821""","""METHADON""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""02_074821""","""MORPHINE""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""02_074821""","""PCP""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0


In [48]:
# Valeur unique du patient
abs_neg = abs_neg.unique(subset=["USUBJID"])
abs_neg

USUBJID,LBTESTCD,LBORRES,LBSTRESC,LBSPEC,VISIT,VISITNUM
str,str,str,str,str,str,i64
"""01_038734""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""02_074821""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""02_014527""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_009835""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_058191""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_078258""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_066150""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""02_065216""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0
"""01_034039""","""AMPHET""","""NEGATIVE""","""NEGATIVE""","""URINE""","""BASELINE, ANY DAY PRIOR TO FIR…",0


Finalement, cela était une mauvaise idée de filtrer selon ces critères, car le résultat final est bien trop bas ! 

Donc nous décidons de sélectionner les sujets présents jusqu'au 13e jour et ayant un test urinaire négatif.

In [68]:
lb_ret.shape # varaible calculé précédamment, dans lequel nous avons filtrer les sujets dont la présence > 12 jour.

(6302, 7)

In [67]:
lb_neg = (
    lb_ret
    .filter(pl.col("LBORRES") == "NEGATIVE"
))

lb_neg

USUBJID,VISIT,VISITNUM,EPOCH,ARMCD,LBORRES,LBSTRESC
str,str,i64,str,str,str,str
"""01_001362""","""FOLLOW-UP 1""",15,"""SCREENING""","""CLON""","""NEGATIVE""","""NEGATIVE"""
"""01_001362""","""FOLLOW-UP 2""",16,"""SCREENING""","""CLON""","""NEGATIVE""","""NEGATIVE"""
"""01_001362""","""FOLLOW-UP 1""",15,"""SCREENING""","""CLON""","""NEGATIVE""","""NEGATIVE"""
"""01_001362""","""FOLLOW-UP 2""",16,"""SCREENING""","""CLON""","""NEGATIVE""","""NEGATIVE"""
"""01_001362""","""FOLLOW-UP 1""",15,"""SCREENING""","""CLON""","""NEGATIVE""","""NEGATIVE"""
…,…,…,…,…,…,…
"""02_099926""","""FOLLOW-UP 2""",16,"""SCREENING""","""BUPNAL""","""NEGATIVE""","""NEGATIVE"""
"""02_099926""","""FOLLOW-UP 3""",17,"""SCREENING""","""BUPNAL""","""NEGATIVE""","""NEGATIVE"""
"""02_099926""","""FOLLOW-UP 3""",17,"""SCREENING""","""BUPNAL""","""NEGATIVE""","""NEGATIVE"""


In [65]:
lb_neg = lb_neg.group_by(
    pl.col('ARMCD'),
    maintain_order=True,
).agg(
    pl.col('USUBJID').unique_counts().len()
)

lb_neg

ARMCD,USUBJID
str,u32
"""CLON""",54
"""BUPNAL""",191
"""SCRFAIL""",1


# Calcul de la proportion de la rétention et d'abstinence

Rétention (%) = Nombre de participants ayant complété l’étude pour un traitement donné / Nombre de participants inclus ​× 100 

Abstinence continue (%) = Nombre de participants avec un test négatif​ selon le traitement / Nombre de participants évalués x 100

In [ ]:
96 / 110  # pourcentage de rétention pour le bras CLON 

In [ ]:
54 / 96  # pourcentage de d'abstinence pour le bras CLON 

In [ ]:
227 / 233  # pourcentage de rétention pour le bras BUP/NA

In [ ]:
191 / 227  # pourcentage d'abstinence pour le bras BUP/NA

Résumons ces résultats dans le table suivant : 

|                    | BUP/NA       | CLON       | Diff   | P |
|--------------------|--------------|------------|--------|---|
| N                  | 233          | 110        | -      | - |
| Rétention          | 227 (97,4 %) | 96 (87,3%) | 10,1%  |  - |- |  |
| Abstinence (+ rét) | 191 (84,1 %) | 54 (56,2%) | 27,9%  | - | - |- ||- |

Outil de construction : https://www.tablesgenerator.com/markdown_tables

Nous avons choisi d’appliquer le test du Khi-deux, conformément aux recommandations méthodologiques utilisées pour l’analyse de cet essai clinique.

# Analyse statistique : Comparaison de l'efficacité BUP/NA vs CLON

Les critères de jugement évalués (rétention et abstinence) sont des **variables binaires** :
- **Rétention** : le patient est présent (succès = 1) ou absent (échec = 0) à VISITNUM > 12
- **Abstinence** : le test urinaire aux opioïdes est négatif (succès = 1) ou positif (échec = 0)

Ces variables peuvent être résumées par des **proportions de succès** dans chaque bras de traitement. L’objectif est donc de comparer des **proportions entre deux groupes indépendants**.

## Choix du test statistique

Pour comparer l’efficacité des deux traitements (**BUP/NA vs CLON**), nous avons choisi d’utiliser **le test du Khi-deux (χ²) d’indépendance**.

Le test du Khi-deux permet de comparer les **distributions de fréquences** (succès / échec) entre deux groupes indépendants. Il est particulièrement adapté à l’analyse de **variables qualitatives binaires**, comme la rétention et l’abstinence.

Ce test permet d’évaluer s’il existe une **association statistiquement significative** entre le type de traitement administré et la probabilité de succès.
<!--  -->
## Condition d'application du test Khi-2
Les conditions d’utilisation du test du Khi-deux sont respectées dans notre étude :
- Variables qualitatives binaires (succès / échec)
- Groupes indépendants (BUP/NA vs CLON)
- Effectifs suffisants dans chaque cellule (effectifs attendus ≥ 5)

En cas d’effectifs insuffisants dans certaines cellules, un **test exact de Fisher** pourrait être envisagé en alternative.

## Taille des échantillons
Dans notre cas, nos échantillons sont suffisamment grands :
- **Rétention** : n₁ = 233 (BUP/NA), n₂ = 110 (CLON)
- **Abstinence** : n₁ = 227 (BUP/NA), n₂ = 96 (CLON)

Ces tailles d’échantillon garantissent une bonne validité asymptotique du test.
## Hypothèses du test t
**Hypothèses du test de Khi-deus :**<br>
Soit :
- p1 ​: probabilité de succès dans le groupe BUP/NA.
- p2 ​: probabilité de succès dans le groupe CLON.

#### Hypothèse nulle (H0)

- H0 : Les proportions sont égales ($p1 = p2$)

L’hypothèse nulle stipule que la probabilité de succès (rétention ou abstinence) est identique dans les deux groupes, c’est-à-dire que le traitement par BUP/NA n’est ni supérieur ni inférieur à la clonidine.

#### Hypothèse alternative 

- H1 : Les proportions sont différentes ($p1 ≠ p2$)

L’hypothèse alternative stipule qu’il existe une différence entre les deux groupes de traitement. Il s’agit d’un **test bilatéral**, sans hypothèse directionnelle a priori.

## Mesure de l'effet et intervalle de confiance

En complément du test Khi-deux, l'effet du traitement sera quantifié par : 
- La différence de proportions : $Δ=p1​−p2$
- Son intervalle de confiances à 95%.<br>
Pour calculer son intervalle, nous appliquons la formule suivant : 

$$
IC = (p_1 - p_2) \pm Z_{1-\alpha/2} \sqrt{\frac{p_1(1-p_1)}{n_1} + \frac{p_2(1-p_2)}{n_2}}
$$



Ces mesures permettent d’évaluer l’ampleur clinique de l’effet observé, au-delà de la seule significativité statistique

## Seuil de signification
Nous nous plaçons à un niveau d'erreur de : **α = 0,05** (5%).<br>

Si la valeur de p issue du test du Khi-deux est inférieure à 0,05, l’hypothèse nulle sera rejetée et la différence observée sera considérée comme **statistiquement significative**.

Sources : 
- https://numiqo.fr/tutorial/chi-square-test
- https://www.jmp.com/fr/statistics-knowledge-portal/chi-square-test
- cours de licence 

# TEST KHI-2

### Test pour la rétention

In [77]:
# Table de contigence : Rétention 
table_ret = np.array([
    [227, 6],    # BUP/NAL: retenus, perdus
    [96, 14]     # CLON: retenus, perdus
])

In [78]:
ret = sp.stats.chi2_contingency(table_ret)
ret

Chi2ContingencyResult(statistic=12.237825229903038, pvalue=0.0004683038058351793, dof=1, expected_freq=array([[219.41399417,  13.58600583],
       [103.58600583,   6.41399417]]))

source : https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html 

- statistic: Plus cette valeur est élevée, plus il y a de différence entre les fréquences observées et attendues, donc plus l’association entre les variables est probable.
- pvalue : C’est la probabilité de trouver une différence aussi extrême (ou plus) si l’hypothèse nulle est vraie. 
- dof : degré de liberté du test.
- expected_freq : ce sont les fréquences attendues sous l’hypothèse d’indépendance.

Nous calculons par la suite manuellement l'intervalle de confiance de la rétention qui encadre l’effet estimé, indique la précision de l’estimation et aide à interpréter sa signification pratique et clinique.

In [ ]:
n_BUP = 233
n_CLON = 110
p_BUP = 227/233
p_CLON = 96/110
diff_ret = p_BUP - p_CLON
diff_ret

In [ ]:
se_ret = np.sqrt((p_BUP*(1-p_BUP)/n_BUP) + (p_CLON*(1-p_CLON)/n_CLON))
se_ret

source : https://sites.google.com/view/aide-python/statistiques/moyennes-et-intervalles-de-confiance-en-python

In [ ]:
ic_ret = (diff_ret - 1.96*se_ret, diff_ret + 1.96*se_ret)
ic_ret 

#### Interprétation :
- χ² = 12.24 → grande différence observée vs attendue (différence de groupe)
- p = 0.000468 → ici p < 0.05 → on rejette l’hypothèse nulle donc il existe une **différence significative** entre nos variables. (la probabilité que cette différence soit due au hasard est très faible)
- - IC 95% : [0.036, 0.167]
  - en % : [3.6%, 16.7%]
  - on peut dire avec 95 % de confiance que BUP/NA améliore la rétention de 3,6 % à 16,7 % par rapport à CLON.

### Test pour l'abstinence

In [72]:
# Table de contigence : Abstinence

table_abs = np.array([
    [191, 36],  # BUP/NA : abstinents, non-abstinents
    [54, 42]    # CLON   : abstinents, non-abstinents
])

In [73]:
abs = sp.stats.chi2_contingency(table_abs)
abs

Chi2ContingencyResult(statistic=27.150290623528015, pvalue=1.8823690027502525e-07, dof=1, expected_freq=array([[172.18266254,  54.81733746],
       [ 72.81733746,  23.18266254]]))

In [74]:
#Calcule de IC95% pour l'abstinence

n2_BUP = 227      # Patients BUP/NAL retenus
n2_CLON = 96      # Patients CLON retenus
p2_BUP = 191/227  # Abstinents BUP/NAL
p2_CLON = 54/96   # Abstinents CLON
diff_abs = p2_BUP - p2_CLON

In [75]:
se_abs = np.sqrt((p2_BUP*(1-p2_BUP)/n2_BUP) + (p2_CLON*(1-p2_CLON)/n2_CLON))
se_abs 

0.05613657834217873

In [76]:
ic_abs = (diff_abs - 1.96*se_abs , diff_abs + 1.96*se_abs )
ic_abs

(0.1688819980792856, 0.38893738518062626)

#### Interprétation : 
- χ² = 27.1 → différence observée vs attendue
- p = 1.88e-07 → ici p < 0.05 → on rejette l’hypothèse nulle donc il existe une **différence très significative** entre nos variables.
- - IC 95% : [0.169, 0.389]
  - en % : [16,9%, 38.9%]
  - on peut dire avec 95 % de confiance que BUP/NA améliore l’abstinence de 16,9% à 38.9% par rapport à CLON.

|                    | BUP/NA       | CLON       | Diff   | p-value  | IC95%           | Significative |
|--------------------|--------------|------------|--------|----------|-----------------|---------------|
| N                  | 233          | 110        | -      | -        | -               | -             |
| Rétention          | 227 (97,4 %) | 96 (87,3%) | 10,1%  | 4,68e-04 | [3,6% ; 16.7%]  | OUI           |
| Abstinence (+ rét) | 191 (84,1 %) | 54 (56,2%) | 27,9%  | 1.88e-07 | [16,9% ; 38,9%] | OUI           |   |

CONCLUSION : Dans les deux cas (rétention et abstinence), nous rejetons H0, ce qui signifie que le traitement au BUP/NA est significativement plus efficace que le traitement CLON.